ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/BackingStore/LMDB.hs

```diff
@@ -0,0 +1,715 @@
+
+-- | @'rangeRead' n db codec ksMay@ performs a range read of @count@ values from
+-- database @db@, starting from some key depending on @ksMay@.
+--
+-- The @codec@ argument defines how to serialise/deserialise keys and values.
+--
+-- A range read can return less than @count@ values if there are not enough
+-- values to read.
+--
+-- Note: See @`RangeQuery`@ for more information about range queries. In
+-- particular, @'rqPrev'@ describes the role of @ksMay@.
+--
+-- What the "first" key in the database is, and more generally in which order
+-- keys are read, depends on the lexographical ordering of the /serialised/
+-- keys. Care should be taken such that the @'Ord'@ instance for @k@ matches the
+-- lexicographical ordering of the serialised keys, or the result of this
+-- function will be unexpected.
+rangeRead ::
+     forall k v mode. Ord k
+  => Int
+  -> LMDBMK k v
+  -> CodecMK k v
+  -> (Maybe :..: KeysMK) k v
+  -> LMDB.Transaction mode (ValuesMK k v)
+rangeRead count dbMK codecMK ksMK =
+    ValuesMK <$> case unComp2 ksMK of
+      Nothing -> runCursorHelper Nothing
+      Just (KeysMK ks) -> case Set.lookupMax ks of
+        Nothing -> pure mempty
+        Just lastExcludedKey ->
+          runCursorHelper $ Just (lastExcludedKey, LMDB.Cursor.Exclusive)
+  where
+    LMDBMK _ db = dbMK
+
+    runCursorHelper ::
+         Maybe (k, LMDB.Cursor.Bound)    -- ^ Lower bound on read range
+      -> LMDB.Transaction mode (Map k v)
+    runCursorHelper lb =
+      Bridge.runCursorAsTransaction'
+        (LMDB.Cursor.cgetMany lb count)
+        db
+        codecMK
+
+initLMDBTable ::
+     LMDBMK   k v
+  -> CodecMK  k v
+  -> ValuesMK k v
+  -> LMDB.Transaction LMDB.ReadWrite (EmptyMK k v)
+initLMDBTable (LMDBMK tblName db) codecMK (ValuesMK utxoVals) =
+    EmptyMK <$ lmdbInitTable
+  where
+    lmdbInitTable  = do
+      isEmpty <- LMDB.null db
+      unless isEmpty $ liftIO . throwIO $ DbErrInitialisingNonEmpty tblName
+      void $ Map.traverseWithKey
+                 (Bridge.put codecMK db)
+                 utxoVals
+
+readLMDBTable ::
+     Ord k
+  => LMDBMK  k v
+  -> CodecMK k v
+  -> KeysMK  k v
+  -> LMDB.Transaction mode (ValuesMK k v)
+readLMDBTable (LMDBMK _ db) codecMK (KeysMK keys) =
+    ValuesMK <$> lmdbReadTable
+  where
+    lmdbReadTable = foldlM' go Map.empty (Set.toList keys)
+      where
+        go m k = Bridge.get codecMK db k <&> \case
+          Nothing -> m
+          Just v  -> Map.insert k v m
+
+writeLMDBTable ::
+     LMDBMK  k v
+  -> CodecMK k v
+  -> DiffMK  k v
+  -> LMDB.Transaction LMDB.ReadWrite (EmptyMK k v)
+writeLMDBTable (LMDBMK _ db) codecMK (DiffMK d) =
+    EmptyMK <$ lmdbWriteTable
+  where
+    lmdbWriteTable = void $ traverseDiffEntryWithKey_ go d
+      where
+        go k de = case de of
+          Delete _v           -> void $ Bridge.delete codecMK db k
+          Insert v            -> Bridge.put codecMK db k v
+          UnsafeAntiDelete _v -> error "Found anti-delete."
```

Comment:
Add a more descriptive error message pointing to the place where this is explained (`diff-containers`?)